The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16851285156/job/47737917992) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/47737917992) [(plain)](https://github.com/rust-lang/rust/commit/21bdfa14cbd99aba173751a2c723fb1e65de401b/checks/47737917992/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [codegen] tests/codegen-llvm/intrinsics/select_unpredictable.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-19/bin/FileCheck" "--input-file" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/intrinsics/select_unpredictable/select_unpredictable.ll" "/checkout/tests/codegen-llvm/intrinsics/select_unpredictable.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
stdout: none
--- stderr -------------------------------
/checkout/tests/codegen-llvm/intrinsics/select_unpredictable.rs:47:12: error: CHECK: expected string not found in input
 // CHECK: select i1 %p, i64 %a, i64 %b, !unpredictable
           ^
/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/intrinsics/select_unpredictable/select_unpredictable.ll:39:30: note: scanning from here
define noundef i64 @test_int2(i1 noundef zeroext %p, i64 noundef %a, i64 noundef %b) unnamed_addr #0 personality ptr @rust_eh_personality {
                             ^
/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/intrinsics/select_unpredictable/select_unpredictable.ll:41:11: note: possible intended match here
 %ret.i = select i1 %p, i64 %a, i64 %b
          ^
/checkout/tests/codegen-llvm/intrinsics/select_unpredictable.rs:54:12: error: CHECK: expected string not found in input
 // CHECK: select i1 %p, {{.*}}, !unpredictable
           ^
/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/intrinsics/select_unpredictable/select_unpredictable.ll:46:32: note: scanning from here
define { i64, i64 } @test_pair2(i1 noundef zeroext %p, i64 noundef %a.0, i64 noundef %a.1, i64 noundef %b.0, i64 noundef %b.1) unnamed_addr #0 personality ptr @rust_eh_personality {
                               ^
/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/intrinsics/select_unpredictable/select_unpredictable.ll:48:29: note: possible intended match here
 %ret.0.i.sroa.speculated = select i1 %p, i64 %a.0, i64 %b.0
                            ^

Input file: /checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/intrinsics/select_unpredictable/select_unpredictable.ll
Check file: /checkout/tests/codegen-llvm/intrinsics/select_unpredictable.rs

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            1: ; ModuleID = 'select_unpredictable.e9f1accfe315d8ff-cgu.0' 
            2: source_filename = "select_unpredictable.e9f1accfe315d8ff-cgu.0" 
            3: target datalayout = "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128-Fn32" 
            4: target triple = "aarch64-unknown-linux-gnu" 
            5:  
            6: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
            7: define noundef i64 @test_int(i1 noundef zeroext %p, i64 noundef %a, i64 noundef %b) unnamed_addr #0 { 
            8: start: 
            9:  %0 = select i1 %p, i64 %a, i64 %b, !unpredictable !2 
           10:  ret i64 %0 
           11: } 
           12:  
           13: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           14: define { i64, i64 } @test_pair(i1 noundef zeroext %p, i64 noundef %a.0, i64 noundef %a.1, i64 noundef %b.0, i64 noundef %b.1) unnamed_addr #0 { 
           15: start: 
           16:  %0 = insertvalue { i64, i64 } poison, i64 %a.0, 0 
           17:  %1 = insertvalue { i64, i64 } %0, i64 %a.1, 1 
           18:  %2 = insertvalue { i64, i64 } poison, i64 %b.0, 0 
           19:  %3 = insertvalue { i64, i64 } %2, i64 %b.1, 1 
           20:  %4 = select i1 %p, { i64, i64 } %1, { i64, i64 } %3, !unpredictable !2 
           21:  ret { i64, i64 } %4 
           22: } 
           23:  
           24: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(read, argmem: readwrite, inaccessiblemem: none) uwtable 
           25: define void @test_struct(ptr dead_on_unwind noalias nocapture noundef writable writeonly sret([800 x i8]) align 8 dereferenceable(800) %_0, i1 noundef zeroext %p, ptr noalias nocapture noundef readonly align 8 dereferenceable(800) %a, ptr noalias nocapture noundef readonly align 8 dereferenceable(800) %b) unnamed_addr #1 { 
           26: start: 
           27:  %0 = select i1 %p, ptr %a, ptr %b, !unpredictable !2 
           28:  tail call void @llvm.memcpy.p0.p0.i64(ptr noundef nonnull align 8 dereferenceable(800) %_0, ptr noundef nonnull align 8 dereferenceable(800) %0, i64 800, i1 false) 
           29:  ret void 
           30: } 
           31:  
           32: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           33: define void @test_zst(i1 noundef zeroext %p) unnamed_addr #0 { 
           34: start: 
           35:  ret void 
           36: } 
           37:  
           38: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           39: define noundef i64 @test_int2(i1 noundef zeroext %p, i64 noundef %a, i64 noundef %b) unnamed_addr #0 personality ptr @rust_eh_personality { 
check:47'0                                  X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           40: start: 
check:47'0     ~~~~~~~
           41:  %ret.i = select i1 %p, i64 %a, i64 %b 
check:47'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:47'1               ?                             possible intended match
           42:  ret i64 %ret.i 
check:47'0     ~~~~~~~~~~~~~~~~
           43: } 
check:47'0     ~~
           44:  
check:47'0     ~
           45: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
check:47'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           46: define { i64, i64 } @test_pair2(i1 noundef zeroext %p, i64 noundef %a.0, i64 noundef %a.1, i64 noundef %b.0, i64 noundef %b.1) unnamed_addr #0 personality ptr @rust_eh_personality { 
check:47'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:54'0                                    X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           47: start: 
check:54'0     ~~~~~~~
           48:  %ret.0.i.sroa.speculated = select i1 %p, i64 %a.0, i64 %b.0 
check:54'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:54'1                                 ?                                 possible intended match
           49:  %ret.1.i.sroa.speculated = select i1 %p, i64 %a.1, i64 %b.1 
check:54'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           50:  %0 = insertvalue { i64, i64 } poison, i64 %ret.0.i.sroa.speculated, 0 
check:54'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           51:  %1 = insertvalue { i64, i64 } %0, i64 %ret.1.i.sroa.speculated, 1 
check:54'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           52:  ret { i64, i64 } %1 
check:54'0     ~~~~~~~~~~~~~~~~~~~~~
           53: } 
check:54'0     ~~
           54:  
check:54'0     ~
           55: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(read, argmem: readwrite, inaccessiblemem: none) uwtable 
check:54'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           56: define void @test_struct2(ptr dead_on_unwind noalias nocapture noundef writable writeonly sret([800 x i8]) align 8 dereferenceable(800) %_0, i1 noundef zeroext %p, ptr noalias nocapture noundef readonly align 8 dereferenceable(800) %a, ptr noalias nocapture noundef readonly align 8 dereferenceable(800) %b) unnamed_addr #1 personality ptr @rust_eh_personality { 
check:54'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
           57: start: 
           58:  %0 = select i1 %p, ptr %a, ptr %b, !unpredictable !2 
           59:  tail call void @llvm.memcpy.p0.p0.i64(ptr noundef nonnull align 8 dereferenceable(800) %_0, ptr noundef nonnull align 8 dereferenceable(800) %0, i64 800, i1 false), !noalias !3 
           60:  ret void 
           61: } 
           62:  
           63: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           64: define void @test_zst2(i1 noundef zeroext %p) unnamed_addr #0 personality ptr @rust_eh_personality { 
           65: start: 
           66:  ret void 
           67: } 
           68:  
           69: ; Function Attrs: nounwind uwtable 
           70: declare noundef range(i32 0, 10) i32 @rust_eh_personality(i32 noundef, i32 noundef, i64 noundef, ptr noundef, ptr noundef) unnamed_addr #2 
           71:  
           72: ; Function Attrs: mustprogress nocallback nofree nounwind willreturn memory(argmem: readwrite) 
           73: declare void @llvm.memcpy.p0.p0.i64(ptr noalias nocapture writeonly, ptr noalias nocapture readonly, i64, i1 immarg) #3 
           74:  
           75: attributes #0 = { mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable "frame-pointer"="non-leaf" "probe-stack"="inline-asm" "target-cpu"="generic" "target-features"="+v8a,+outline-atomics" } 
           76: attributes #1 = { mustprogress nofree norecurse nosync nounwind willreturn memory(read, argmem: readwrite, inaccessiblemem: none) uwtable "frame-pointer"="non-leaf" "probe-stack"="inline-asm" "target-cpu"="generic" "target-features"="+v8a,+outline-atomics" } 
           77: attributes #2 = { nounwind uwtable "frame-pointer"="non-leaf" "probe-stack"="inline-asm" "target-cpu"="generic" "target-features"="+v8a,+outline-atomics" } 
           78: attributes #3 = { mustprogress nocallback nofree nounwind willreturn memory(argmem: readwrite) } 
           79:  
           80: !llvm.module.flags = !{!0} 
           81: !llvm.ident = !{!1} 
           82:  
           83: !0 = !{i32 8, !"PIC Level", i32 2} 
           84: !1 = !{!"rustc version 1.91.0-nightly (00fc429ff 2025-08-09)"} 
           85: !2 = !{} 
           86: !3 = !{!4, !6} 
           87: !4 = distinct !{!4, !5, !"_ZN4core4hint20select_unpredictable17h598d305da1153d4fE: %true_val"} 
           88: !5 = distinct !{!5, !"_ZN4core4hint20select_unpredictable17h598d305da1153d4fE"} 
           89: !6 = distinct !{!6, !5, !"_ZN4core4hint20select_unpredictable17h598d305da1153d4fE: %false_val"} 
>>>>>>
------------------------------------------



```

</details>

--- a/library/core/src/hint.rs
+++ b/library/core/src/hint.rs
@@ -786,16 +786,42 @@ pub fn select_unpredictable<T>(condition: bool, true_val: T, false_val: T) -> T 
     // Change this to use ManuallyDrop instead.
     let mut true_val = MaybeUninit::new(true_val);
     let mut false_val = MaybeUninit::new(false_val);
+
+    struct DropOnPanic<T> {
+        inner: *mut MaybeUninit<T>,
+    }
+
+    impl<T> Drop for DropOnPanic<T> {
+        fn drop(&mut self) {
+            // SAFETY: Must be guaranteed on construction of local type `DropOnPanic`.
+            unsafe { (*self.inner).assume_init_drop() }
+        }
+    }
+
+    let true_ptr = (&mut true_val) as *mut _;
+    let false_ptr = (&mut false_val) as *mut _;
+
     // SAFETY: The value that is not selected is dropped, and the selected one
     // is returned. This is necessary because the intrinsic doesn't drop the
     // value that is  not selected.
     unsafe {
         // Extract the selected value first, ensure it is dropped as well if dropping the unselected
         // value panics.
-        let ret = crate::intrinsics::select_unpredictable(condition, &true_val, &false_val)
-            .assume_init_read();
-        crate::intrinsics::select_unpredictable(!condition, &mut true_val, &mut false_val)
-            .assume_init_drop();
-        ret
+        let (guard, drop) = crate::intrinsics::select_unpredictable(
+            condition,
+            (true_ptr, false_ptr),
+            (false_ptr, true_ptr),
+        );
+
+        // SAFETY: both pointers are to valid `MaybeUninit`, in both variants they do not alias but
+        // the two arguments we have selected from did alias each other.
+        let guard = DropOnPanic { inner: guard };
+        (*drop).assume_init_drop();
+        crate::mem::forget(guard);
+
+        // Note that it is important to use the values here. Reading from the pointer we got makes
+        // LLVM forget the !unpredictable annotation sometimes (in tests, integer sized values in
+        // particular seemed to confuse it, also observed in llvm/llvm-project #82340).
+        crate::intrinsics::select_unpredictable(condition, true_val, false_val).assume_init()
     }
 }

